File: backend/blog_api/views/posts.py
from rest_framework import viewsets, filters, status
from rest_framework.decorators import action
from rest_framework.response import Response
from rest_framework.permissions import IsAuthenticatedOrReadOnly
from django_filters.rest_framework import DjangoFilterBackend
from django.db import models
from ..models import Post
from ..serializers.posts import PostListSerializer, PostDetailSerializer, PostCreateUpdateSerializer
from ..permissions import IsAuthor, IsOwnerOrReadOnly


class PostViewSet(viewsets.ModelViewSet):
    """
    ViewSet for Post model.
    Provides CRUD operations, filtering, and search.
    """
    queryset = Post.objects.select_related('author', 'category').all()
    permission_classes = [IsAuthenticatedOrReadOnly, IsAuthor, IsOwnerOrReadOnly]
    filter_backends = [DjangoFilterBackend, filters.SearchFilter, filters.OrderingFilter]
    filterset_fields = ['status', 'category__slug', 'author__username']
    search_fields = ['title', 'content', 'excerpt']
    ordering_fields = ['created_at', 'updated_at', 'title']
    ordering = ['-created_at']
    lookup_field = 'slug'
    
    def get_serializer_class(self):
        if self.action == 'list':
            return PostListSerializer
        elif self.action in ['create', 'update', 'partial_update']:
            return PostCreateUpdateSerializer
        return PostDetailSerializer
    
    def get_queryset(self):
        queryset = super().get_queryset()
        
        # Non-authenticated users and non-authors only see published posts
        if not self.request.user.is_authenticated or not self.request.user.is_author:
            queryset = queryset.filter(status='published')
        # Authors see their own posts (all statuses) + published posts from others
        elif self.request.user.is_author:
            queryset = queryset.filter(
                models.Q(author=self.request.user) | models.Q(status='published')
            )
        
        return queryset
    
    def perform_create(self, serializer):
        serializer.save(author=self.request.user)
    
    @action(detail=False, methods=['get'], url_path='category/(?P<category_slug>[^/.]+)')
    def by_category(self, request, category_slug=None):
        """
        Get posts by category slug.
        GET /api/v1/posts/category/{slug}/
        """
        posts = self.get_queryset().filter(category__slug=category_slug, status='published')
        page = self.paginate_queryset(posts)
        if page is not None:
            serializer = PostListSerializer(page, many=True)
            return self.get_paginated_response(serializer.data)
        serializer = PostListSerializer(posts, many=True)
        return Response(serializer.data)
    
    @action(detail=False, methods=['get'], url_path='author/(?P<username>[^/.]+)')
    def by_author(self, request, username=None):
        """
        Get posts by author username.
        GET /api/v1/posts/author/{username}/
        """
        posts = self.get_queryset().filter(author__username=username, status='published')
        page = self.paginate_queryset(posts)
        if page is not None:
            serializer = PostListSerializer(page, many=True)
            return self.get_paginated_response(serializer.data)
        serializer = PostListSerializer(posts, many=True)
        return Response(serializer.data)
    
    @action(detail=False, methods=['get'], url_path='my-posts')
    def my_posts(self, request):
        """
        Get current user's own posts (all statuses).
        GET /api/v1/posts/my-posts/
        """
        if not request.user.is_authenticated:
            return Response({"detail": "Authentication required"}, status=status.HTTP_401_UNAUTHORIZED)
        
        posts = Post.objects.filter(author=request.user).select_related('author', 'category').order_by('-created_at')
        page = self.paginate_queryset(posts)
        if page is not None:
            serializer = PostListSerializer(page, many=True)
            return self.get_paginated_response(serializer.data)
        serializer = PostListSerializer(posts, many=True)
        return Response(serializer.data)

File: lib/api/posts.ts
// Posts API service
// Connected to Django REST API

import { apiClient } from "./client"

interface Author {
    id: number
    username: string
    email: string
    is_author: boolean
}

interface Category {
    id: number
    name: string
    slug: string
}

interface Post {
    id: number
    title: string
    slug: string
    content: string
    excerpt: string
    status: "draft" | "published"
    category: Category
    author: Author
    image: string | null
    created_at: string
    updated_at: string
    comments_count?: number
}

interface PostListResponse {
    count: number
    next: string | null
    previous: string | null
    results: Post[]
}

interface CreatePostData {
    title: string
    content: string
    excerpt?: string
    category_id: number
    image?: string
    status: "draft" | "published"
}

export const postsApi = {
    // Get all posts (with pagination)
    async getAllPosts(page: number = 1): Promise<PostListResponse> {
        return apiClient.get<PostListResponse>(`/posts/?page=${page}`)
    },

    // Get single post by slug
    async getPostBySlug(slug: string): Promise<Post> {
        return apiClient.get<Post>(`/posts/${slug}/`)
    },

    // Get posts by category
    async getPostsByCategory(categorySlug: string, page: number = 1): Promise<PostListResponse> {
        return apiClient.get<PostListResponse>(`/posts/category/${categorySlug}/?page=${page}`)
    },

    // Get posts by author
    async getPostsByAuthor(username: string, page: number = 1): Promise<PostListResponse> {
        return apiClient.get<PostListResponse>(`/posts/author/${username}/?page=${page}`)
    },

    // Search posts
    async searchPosts(query: string, page: number = 1): Promise<PostListResponse> {
        return apiClient.get<PostListResponse>(`/posts/?search=${encodeURIComponent(query)}&page=${page}`)
    },

    // Create new post (author only)
    async createPost(data: CreatePostData): Promise<Post> {
        return apiClient.post<Post>("/posts/", data)
    },

    // Update post (owner only)
    async updatePost(slug: string, data: Partial<CreatePostData>): Promise<Post> {
        return apiClient.put<Post>(`/posts/${slug}/`, data)
    },

    // Delete post (owner only)
    async deletePost(slug: string): Promise<void> {
        return apiClient.delete<void>(`/posts/${slug}/`)
    },

    // Get user's own posts (drafts + published)
    async getMyPosts(page: number = 1): Promise<PostListResponse> {
        return apiClient.get<PostListResponse>(`/posts/my-posts/?page=${page}`)
    },
}
